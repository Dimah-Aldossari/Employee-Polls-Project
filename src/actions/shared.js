// actions/shared.js

import { setUsers } from "./usersSlice";
import { setQuestions } from "./questionsSlice";
import { getInitialData } from "../api";

export function handleInitialData() {
    return (dispatch) => {
        return getInitialData().then(({ users, questions }) => {
            dispatch(setUsers(users));        // Corrected usage
            dispatch(setQuestions(questions)); // Corrected usage
        });
    };
}
