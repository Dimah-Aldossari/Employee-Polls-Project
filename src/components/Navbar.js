import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import { _getUsers } from "../_DATA";
import { login, logout } from "../actions/Creat_Users";

const Navbar = () => {
    const dispatch = useDispatch();

    const isAuthenticated = useSelector(
        (state) => state.authSlice.isAuthenticated
    );
    const user = useSelector((state) => state.authSlice.user);
    const navigate = useNavigate();
    const [loginData, setLoginData] = useState({
        username: "",
        password: "",
    });

    const handleLogin = async () => {
        try {
            const users = await _getUsers();
            const { username, password } = loginData;

            if (users[username] && users[username].password === password) {
                dispatch(login(users[username]));
                navigate("/addQuestion");
            } else {
                console.log("Invalid username or password");
            }
        } catch (error) {
            console.error("Error logging in:", error);
        }
    };

    const handleLogout = () => {
        dispatch(logout());
        navigate("/");

    };

    const goToAddNewQuestion = () => {
        navigate("/addQuestion");
    };

    const goToHome = () => {
        navigate("/Home");
    };

    const goToLeaderBoard = () => {
        navigate("/LeaderBoard");
    };

    return (
        <nav>
            <ul>
                {isAuthenticated ? (
                    <>
                        <div className="welcome" data-testid="user-information">
                            <span className="username">welcome: {user.name}</span>

                        </div>
                        <div className="navbar">

                            <div className="nav-buttons">
                                <button className="nav-button home" onClick={goToHome}>Home</button>
                                <button className="nav-button leaderboard-button" data-testid="heading" onClick={goToLeaderBoard}>LeaderBoard</button>

                                <button className="nav-button add-pool" onClick={goToAddNewQuestion}>Add New Pool</button>
                                <button className="nav-button logout" onClick={handleLogout} >Logout</button>
                            </div>
                        </div>
                    </>
                ) : (
                    <li>
                        <input
                            className="input"
                            type="text"
                            placeholder="Username"
                            value={loginData.username}
                            onChange={(e) =>
                                setLoginData({ ...loginData, username: e.target.value })
                            }
                        />
                        <input
                            className="input"
                            type="password"
                            placeholder="Password"
                            value={loginData.password}
                            onChange={(e) =>
                                setLoginData({ ...loginData, password: e.target.value })
                            }
                        />
                        <br></br>
                        <div className="btn-box" >
                            <button className="btn" onClick={handleLogin} data-testid="login-heading">Login</button>

                        </div>
                    </li>
                )}
            </ul>
        </nav>
    );
};

export default Navbar;
