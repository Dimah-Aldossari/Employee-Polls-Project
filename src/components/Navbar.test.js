import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "../Store";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Navbar from "./Navbar";
import { login } from "../actions/Creat_Users"; // Update the import

describe("Navbar", () => {
    it("should render the component", () => {
        // Dispatch login action
        store.dispatch(login({ id: "sarahedo", name: "Sarah Edo" })); // Use the login action creator

        // Render the Navbar component
        const view = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Navbar />
                </BrowserRouter>
            </Provider>
        );

        // Assert that the component renders correctly
        expect(view).toBeDefined();
        expect(view).toMatchSnapshot();
    });

    it("should display username of logged-in user", () => {
        // Dispatch login action
        store.dispatch(login({ id: "sarahedo", name: "Sarah Edo" })); // Use the login action creator

        // Render the Navbar component
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Navbar />
                </BrowserRouter>
            </Provider>
        );

        // Find the element with data-testid="user-information"
        const userSpanElement = screen.getByTestId("user-information");

        // Assert that the content of the element is as expected
        expect(userSpanElement.textContent).toBe("welcome: Sarah Edo");
    });
});
