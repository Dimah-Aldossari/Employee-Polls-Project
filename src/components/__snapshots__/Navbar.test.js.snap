// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav>
        <ul>
          <div
            class="welcome"
            data-testid="user-information"
          >
            <span
              class="username"
            >
              welcome: 
              Sarah Edo
            </span>
          </div>
          <div
            class="navbar"
          >
            <div
              class="nav-buttons"
            >
              <button
                class="nav-button home"
              >
                Home
              </button>
              <button
                class="nav-button leaderboard-button"
                data-testid="heading"
              >
                LeaderBoard
              </button>
              <button
                class="nav-button add-pool"
              >
                Add New Pool
              </button>
              <button
                class="nav-button logout"
              >
                Logout
              </button>
            </div>
          </div>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav>
      <ul>
        <div
          class="welcome"
          data-testid="user-information"
        >
          <span
            class="username"
          >
            welcome: 
            Sarah Edo
          </span>
        </div>
        <div
          class="navbar"
        >
          <div
            class="nav-buttons"
          >
            <button
              class="nav-button home"
            >
              Home
            </button>
            <button
              class="nav-button leaderboard-button"
              data-testid="heading"
            >
              LeaderBoard
            </button>
            <button
              class="nav-button add-pool"
            >
              Add New Pool
            </button>
            <button
              class="nav-button logout"
            >
              Logout
            </button>
          </div>
        </div>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
